# Data Analytics — Exercise 7.2: Pandas Essentials
## Cris | Date: 2026-05-21

This notebook uses the Pandas library to load, inspect, filter, and summarise employee sales data for RetailHub Inc. It covers DataFrame creation, Boolean filtering, calculated columns, apply() with a lambda, and GroupBy aggregation.

## 1. Import Pandas and Create the DataFrame
We import Pandas and build a DataFrame from a dictionary of employee records.

In [6]:
import pandas as pd

In [7]:
data = {
    'name'   : ['Alice', 'Ben', 'Carmen', 'Dev',
                'Eve', 'Frank', 'Grace', 'Hiro'],
    'city'   : ['London', 'Manchester', 'London', 'Leeds',
                'Birmingham', 'Manchester', 'Leeds', 'London'],
    'sales'  : [4200.0, 3100.0, 5050.0, 2800.0,
                1950.0, 3750.0, 4600.0, 3300.0],
    'months' : [12, 6, 24, 3, 18, 9, 15, 7]
}

df = pd.DataFrame(data)

In [8]:
print('── Employee DataFrame ──────────────────────')
print(df)
print(f'\nShape  : {df.shape}')
print(f'dtypes :\n{df.dtypes}')
print('\ndf.info():')
df.info()

── Employee DataFrame ──────────────────────
     name        city   sales  months
0   Alice      London  4200.0      12
1     Ben  Manchester  3100.0       6
2  Carmen      London  5050.0      24
3     Dev       Leeds  2800.0       3
4     Eve  Birmingham  1950.0      18
5   Frank  Manchester  3750.0       9
6   Grace       Leeds  4600.0      15
7    Hiro      London  3300.0       7

Shape  : (8, 4)
dtypes :
name       object
city       object
sales     float64
months      int64
dtype: object

df.info():
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   name    8 non-null      object 
 1   city    8 non-null      object 
 2   sales   8 non-null      float64
 3   months  8 non-null      int64  
dtypes: float64(1), int64(1), object(2)
memory usage: 388.0+ bytes


## 2. Inspect and Describe the Data
We use Pandas methods to preview the data and compute summary statistics.

In [9]:
print('First 3 rows:\n', df.head(3))
print('\nLast 3 rows:\n', df.tail(3))
print('\nSummary statistics:\n', df.describe())

First 3 rows:
      name        city   sales  months
0   Alice      London  4200.0      12
1     Ben  Manchester  3100.0       6
2  Carmen      London  5050.0      24

Last 3 rows:
     name        city   sales  months
5  Frank  Manchester  3750.0       9
6  Grace       Leeds  4600.0      15
7   Hiro      London  3300.0       7

Summary statistics:
              sales     months
count     8.000000   8.000000
mean   3593.750000  11.750000
std    1013.282326   6.964194
min    1950.000000   3.000000
25%    3025.000000   6.750000
50%    3525.000000  10.500000
75%    4300.000000  15.750000
max    5050.000000  24.000000


In [10]:
print(f'\nMean sales : £{df["sales"].mean():>10,.2f}')
print(f'Max sales  : £{df["sales"].max():>10,.2f}')
print(f'Min sales  : £{df["sales"].min():>10,.2f}')
print(f'Total sales: £{df["sales"].sum():>10,.2f}')


Mean sales : £  3,593.75
Max sales  : £  5,050.00
Min sales  : £  1,950.00
Total sales: £ 28,750.00


## 3. Select, Filter, and Sort
Pandas makes it easy to slice columns, filter rows, and sort results.

In [11]:
print('Name & Sales columns:\n', df[['name', 'sales']])

Name & Sales columns:
      name   sales
0   Alice  4200.0
1     Ben  3100.0
2  Carmen  5050.0
3     Dev  2800.0
4     Eve  1950.0
5   Frank  3750.0
6   Grace  4600.0
7    Hiro  3300.0


In [12]:
high_sales = df[df['sales'] > 3500.0]
print('\nSales > £3500:\n', high_sales)


Sales > £3500:
      name        city   sales  months
0   Alice      London  4200.0      12
2  Carmen      London  5050.0      24
5   Frank  Manchester  3750.0       9
6   Grace       Leeds  4600.0      15


In [13]:
df_sorted = df.sort_values('sales', ascending=False).reset_index(drop=True)
print('\nSorted by sales (highest first):\n', df_sorted)


Sorted by sales (highest first):
      name        city   sales  months
0  Carmen      London  5050.0      24
1   Grace       Leeds  4600.0      15
2   Alice      London  4200.0      12
3   Frank  Manchester  3750.0       9
4    Hiro      London  3300.0       7
5     Ben  Manchester  3100.0       6
6     Dev       Leeds  2800.0       3
7     Eve  Birmingham  1950.0      18


In [14]:
print('\nTop seller record:\n', df_sorted.iloc[0])


Top seller record:
 name      Carmen
city      London
sales     5050.0
months        24
Name: 0, dtype: object


## 4. Calculated Column and Conditional Rating
We add commission (8% of sales) and a performance rating using apply().

In [4]:
# Requirement 5 — commission and rating columns

# Add commission column — element-wise (no loop needed)
df['commission'] = df['sales'] * 0.08

# Add rating using apply() with a lambda
df['rating'] = df['sales'].apply(
    lambda x: 'Star' if x >= 4000
    else ('Good' if x >= 2500 else 'Developing')
)

print('Updated DataFrame:\n', df)

# Print each employee's commission using f-string + iterrows()
print('\n── Commission per employee ──────────────')
for _, row in df.iterrows():
    print(f'{row["name"]:<8} | £{row["commission"]:>8,.2f} | {row["rating"]}')

Updated DataFrame:
      name        city   sales  months  commission      rating
0   Alice      London  4200.0      12       336.0        Star
1     Ben  Manchester  3100.0       6       248.0        Good
2  Carmen      London  5050.0      24       404.0        Star
3     Dev       Leeds  2800.0       3       224.0        Good
4     Eve  Birmingham  1950.0      18       156.0  Developing
5   Frank  Manchester  3750.0       9       300.0        Good
6   Grace       Leeds  4600.0      15       368.0        Star
7    Hiro      London  3300.0       7       264.0        Good

── Commission per employee ──────────────
Alice    | £  336.00 | Star
Ben      | £  248.00 | Good
Carmen   | £  404.00 | Star
Dev      | £  224.00 | Good
Eve      | £  156.00 | Developing
Frank    | £  300.00 | Good
Grace    | £  368.00 | Star
Hiro     | £  264.00 | Good


## 5. GroupBy — Sales by City
GroupBy lets us aggregate data by a category, revealing city-level performance.

In [5]:
# Requirement 6 — GroupBy aggregation

# Total sales per city
city_totals = df.groupby('city')['sales'].sum()
print('── Total Sales by City ──────────────────')
print(city_totals)

# Mean sales per city
city_means = df.groupby('city')['sales'].mean()
print('\n── Mean Sales by City ───────────────────')
print(city_means)

# Full aggregation — sum, mean, count per city
city_agg = df.groupby('city').agg({'sales': ['sum', 'mean', 'count']})
print('\n── Full City Summary ────────────────────')
print(city_agg)

# Find the top city
top_city = city_totals.idxmax()
print(f'\nTop city: {top_city}   Total: £{city_totals[top_city]:,.2f}')

── Total Sales by City ──────────────────
city
Birmingham     1950.0
Leeds          7400.0
London        12550.0
Manchester     6850.0
Name: sales, dtype: float64

── Mean Sales by City ───────────────────
city
Birmingham    1950.000000
Leeds         3700.000000
London        4183.333333
Manchester    3425.000000
Name: sales, dtype: float64

── Full City Summary ────────────────────
              sales                   
                sum         mean count
city                                  
Birmingham   1950.0  1950.000000     1
Leeds        7400.0  3700.000000     2
London      12550.0  4183.333333     3
Manchester   6850.0  3425.000000     2

Top city: London   Total: £12,550.00


## Summary
- London is the top-performing city with a combined total of 12,550 from 3 employees.
- 3 of 8 employees achieved Star rating; only Eve (Birmingham) is Developing.
- The team's average monthly sales of 3,593.75 and top performer Carmen means London should be prioritised for investment in the next quarter.